Refactor/sempahore abi to utils

Dozie2001 opened the pull request:
<!-- Please refer to our CONTRIBUTING documentation for any questions on submitting a pull request. -->
<!-- Provide a general summary of your changes in the Title above. -->

## Description

This PR adds a script that automates moving the ABI to the  ```utils``` package <br> 
Additionally, Semaphore ABI will be able to import it directly from the ```@semaphore-protocol/utils``` package

<!-- Describe your changes in detail. -->
<!-- You may want to answer some of the following questions: -->
<!-- What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...) -->
<!-- What is the current behavior?** (You can also link to an open issue here) -->
<!-- What is the new behavior (if this is a feature change)? -->
<!-- Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?) -->

## Related Issue(s)
Closes #817 
<!-- This project accepts pull requests related to open issues. -->
<!-- If suggesting a new feature or change, please discuss it in an issue first. -->
<!-- If fixing a bug, there should be an issue describing it with steps to reproduce. -->
<!-- Please link to the issue(s) here -->

<!-- Closes # -->
<!-- Fixes # -->

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->
<!-- Feel free to remove this section if you will not use it. -->

## Checklist

<!-- Please check if the PR fulfills these requirements. -->

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] My changes generate no new warnings
-   [x] I have run `yarn format` and `yarn lint` without getting any errors
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes

Dozie2001 committed:
ci(contracts): a script that moves the abi.json file created to utils folder

A script could be run immediately after compilation to copy the file
artifacts/contracts/Semaphore.sol/Semaphore.json to packages/utils/src/semaphore-abi.json. Then all
packages/apps that need the interface can import it from @semaphore-protocol/utils.

re #817

Dozie2001 committed:
refactor(utils): allows packages to import abi from @semaphore-protocol/utils

re #817

cedoor commented on package.json:
```diff
@@ -10,7 +10,8 @@
         "build:libraries": "yarn workspaces foreach -Apt --no-private run build",
         "build:subgraph": "yarn workspace semaphore-subgraph build",
-        "compile:contracts": "yarn workspace semaphore-contracts compile",
+        "copy:abi": "cp  packages/contracts/artifacts/contracts/Semaphore.sol/Semaphore.json packages/utils/src/semaphore-abi.json",
```
`cp` will only work on Unix-like systems. A Node package alternative like [`ncp`](https://www.npmjs.com/package/ncp) should be better here.

I'd also suggest moving this script (`copy:abi`) to the `packages/contracts/package.json` file directly, otherwise if I run `yarn compile` from `packages/contracts` this script won't be executed.

cedoor replied:
`semaphore-abi.json` actually includes bytecode and other data. `semaphore-interface.json` should be more representative.

cedoor commented on packages/utils/src/constants.ts:
```diff
@@ -1,5 +1,10 @@
 export const MAX_DEPTH = 32
+
+// Semaphore ABI
+export const SemaphoreABI = abi.abi
```
Related to the comments above. Here we could export abi and bytecode. Something like:

```ts
export const SemaphoreABI = interface.abi
export const SemaphoreBytecode = interface.bytecode
```

No need to write comments if the variables are self-explanatory.

Dozie2001 replied:
Do you need to export Bytecode separaretly? or do you just want the option of doing so?

cedoor replied:
What do you mean by separately? I think two exports like the ones above should be fine. Wdyt?

Dozie2001 replied:
Hmm no worrries will work on that now

cedoor reviewed (changes requested):
Thanks @Dozie2001 :) Left some comments!

Dozie2001 committed:
refactor(cli-template-monorepo-ethers): installs ncp and exports Semaphore Abi byte code

re #817

Dozie2001 committed:
refactor(contracts): added a script that automates movement of semaphore-abi.json

cedoor commented on packages/contracts/package.json:
```diff
@@ -39,6 +41,7 @@
     "dependencies": {
-        "@zk-kit/lean-imt.sol": "2.0.0"
+        "@zk-kit/lean-imt.sol": "2.0.0",
+        "ncp": "^2.0.0"
```
`ncp` should be a dev dependency

cedoor commented on packages/utils/src/constants.ts:
```diff
@@ -1,5 +1,10 @@
+import abi from "./semaphore-interface.json"
```
`_interface` as a name should be ok here, as it's not the abi actually.

Dozie2001 committed:
chore(cli): changed the dependency and moved to ncp for copying semaphore abi

Changed the importation of abi in utils to _interface and moved the ncp to a dev depency

Dozie2001 committed:
chore(utils): changed ab to _interface and update to package.json

cedoor commented on packages/contracts/package.json:
```diff
@@ -3,6 +3,7 @@
     "private": true,
     "scripts": {
         "start": "hardhat node",
+        "copy:abi": "ncp  artifacts/contracts/Semaphore.sol/Semaphore.json ../utils/src/semaphore-abi.json",
```
This script here should be enough. No need to have one in the main `package.json` file. Just make sure to add a `&& yarn copy:abi` after `hardhat compile` here.

cedoor reviewed (changes requested):
Not related to this PR, it's an issue we're going to solve in another one. But could you remove the files in the cli-templates folder from this PR?

Dozie2001 committed:
chore(cli): deleted Cli-template file and made a document change

delted some cli-template file, changed yarn commit to git commit and fixed the yarn compile  script

cedoor reviewed (approved):
LGTM @Dozie2001, thank you vm :)
